src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/RetriableX509ChainBuildPolicy.cs

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+
+namespace NuGet.Packaging.Signing
+{
+    internal sealed class RetriableX509ChainBuildPolicy : IX509ChainBuildPolicy
+    {
+        // These properties are non-private only to facilitate testing.
+        internal int RetryCount { get; }
+        internal TimeSpan SleepInterval { get; }
+
+        internal RetriableX509ChainBuildPolicy(int retryCount, TimeSpan sleepInterval)
+        {
+            if (retryCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(retryCount));
+            }
+
+            if (sleepInterval < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(sleepInterval));
+            }
+
+            RetryCount = retryCount;
+            SleepInterval = sleepInterval;
+        }
+
+        public bool Build(X509Chain chain, X509Certificate2 certificate)
+        {
+            if (chain is null)
+            {
+                throw new ArgumentNullException(nameof(chain));
+            }
+
+            if (certificate is null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+
+            bool wasBuilt = chain.Build(certificate);
+
+            for (var i = 0; i < RetryCount; ++i)
+            {
+                if (wasBuilt)
+                {
+                    break;
+                }
+
+                bool hasUntrustedRoot = false;
+
+                foreach (X509ChainStatus chainStatus in chain.ChainStatus)
+                {
+                    if (chainStatus.Status.HasFlag(X509ChainStatusFlags.UntrustedRoot))
+                    {
+                        hasUntrustedRoot = true;
+                        break;
+                    }
+                }
+
+                if (!hasUntrustedRoot)
+                {
+                    break;
+                }
+
+                Thread.Sleep(SleepInterval);
+
+                wasBuilt = chain.Build(certificate);
```

Comment:
`RetriableX509ChainBuildPolicy`'s constructor could accept an `IX509ChainBuildPolicy` whose `Build` method you could call here. This has a few benefits:

1. The unit tests for `RetriableX509ChainBuildPolicy` wouldn't need to do any chain building as the inner `IX509ChainBuildPolicy` can just be a mock.
2. In the future it would allow us to compose multiple `IX509ChainBuildPolicy` together, similar to the middleware pattern in ASP.NET Core.

Reply:
Or `DefaultX509ChainBuildPolicy` could not be `sealed`, `bool Build(X509Chain, X509Certificate2)` could be virtual and `RetriableX509ChainBuildPolicy` could inherit `DefaultX509ChainBuildPolicy`, and add the retries around the base `Build()`?

Reply:
![image](https://user-images.githubusercontent.com/737941/133188242-fef32e1b-a2a8-45a4-b7b1-868855c6e974.png)

Reply:
If we use composition, we might as well have a single class that the factory composes instead, right?

Reply:
That'd work too, but I'm not sure I see the benefits off-hand. It may make unit testing a little harder.